Building on Raspberry Pi

I have tried to get the library to build on my Raspberry Pi 2, i.e. on ARM architecture, with little success so far. 

For starters, the `-m32`/`-m64` does not work there, and after removal of them from Makefile it at least builds. Running the `make test` command yields the following output:

```
pi@raspberrypi:~/Repos/Metawear-CppAPI $ make test
python3 -m unittest discover -s test
terminate called after throwing an instance of 'std::out_of_range'
  what():  _Map_base::at
Makefile:117: recipe for target 'test' failed
make: *** [test] Aborted
```
I have tried several other flags (The Rasbian Jessie image has `gcc (Raspbian 4.9.2-10) 4.9.2` as default so I was looking through the [manual](https://gcc.gnu.org/onlinedocs/gcc-4.9.2/gcc/ARM-Options.html#ARM-Options)), but also without success. Have tried building with Clang as well, but the apt-get package of it is this:
```
Raspbian clang version 3.5.0-10+rpi1 (tags/RELEASE_350/final) (based on LLVM 3.5.0)
Target: arm-unknown-linux-gnueabihf
Thread model: posix
```

Do you have any idea as to what might be done to get it to run? This error is similar to the one in #2, is the cause similar as well do you think?

```
pi@raspberrypi:~/Metawear-CppAPI $ make test
python3 -m unittest discover -s test -v
test_acc_bma255_disable_acceleration_sampling (testaccelerometer.TestGenericAccelerometerBma255) ... terminate called after throwing an instance of 'std::out_of_range'
  what():  _Map_base::at
Makefile:97: recipe for target 'test' failed
make: *** [test] Aborted
```

I haven't tried other builds, are you saying this works with Ubuntu on the Pi?

Added print statement as you instructed, but it never arrives at it apparently since nothing is printed out. Hunted it down to [testaccelerometer.py](https://github.com/hbldh/Metawear-CppAPI/blob/master/test/testaccelerometer.py#L189) where I could surround this call with print statements and only have the one prior to the call printed.

I removed that function to see what happened with the tests then and got the following output:
```
pi@raspberrypi:~/Repos/Metawear-CppAPI $ make test
python3 -m unittest discover -s test -v
test_disable_acceleration_sampling (testaccelerometer_bma255.TestAccBma255Config) ... [3, 2, 0, 1]
ok
test_enable_acceleration_sampling (testaccelerometer_bma255.TestAccBma255Config) ... [3, 2, 1, 0]
ok
test_set_odr (testaccelerometer_bma255.TestAccBma255Config) ... terminate called after throwing an instance of 'std::out_of_range'
  what():  _Map_base::at
Makefile:117: recipe for target 'test' failed
make: *** [test] Aborted

```

So, using the generic function `mbl_mw_acc_disable_acceleration_sampling`, it crashes, but when using the model specific call `mbl_mw_acc_bosch_disable_acceleration_sampling`, it works fine. Does this indicate that something with accessing the sent in board's details creates the segfault?

Results:
```
python3 -m unittest discover -s test
setUp: Pre-Init /** Before L209 **/
setUp: Post-Init /** After L209 **/
terminate called after throwing an instance of 'std::out_of_range'
  what():  _Map_base::at
Makefile:99: recipe for target 'test' failed
make: *** [test] Aborted
```
The print statements surrounding L209 gets executed, but nothing is executed in the `initialized` function in L221.

Could not refrain from some more testing...

I believe the fact that `long` on the [Raspberry is a 32-bit value](http://www.raspberry-projects.com/pi/programming-in-c/memory/variables) is the culprit.

Changing [Epoch and GattCharacterstics](https://github.com/mbientlab/Metawear-CppAPI/blob/master/wrapper/python/mbientlab/metawear/core.py#L40-L55) values to be of C type `longlong` instead makes all but one unit test pass: [`test_unknown_entry` in `testlogging.py`](https://github.com/mbientlab/Metawear-CppAPI/blob/master/test/testlogging.py#L119-L124). It creates a segmentation fault:

```
test_unknown_entry (testlogging.TestLogDownload) ... 
Char: 0x2a2600001000 0x800000805f9b34fb
Char: 0x2a2400001000 0x800000805f9b34fb
initialized: Board: 26701304 status: 0
Makefile:107: recipe for target 'test' failed
make: *** [test] Segmentation fault
```
Have you got any idea as to why that particular test fails? 

How do you think this discovery can be integrated in MetaWear-CppAPI? The use of `c_longlong` instead of `c_long` in the Python wrapper should not create any problems on x86 and x86_64 machines, am I right?

--- a/Makefile
+++ b/Makefile
@@ -39,8 +39,9 @@ ifeq ($(MACHINE),x86)
 else ifeq ($(MACHINE),x64)
     CXXFLAGS+=-m64
     LD_FLAGS+=-m64
+else ifeq ($(MACHINE),arm)
 else
-    $(error Unrecognized "MACHINE" value, use 'x86' or 'x64')
+    $(error Unrecognized "MACHINE" value, use 'x86', 'x64' or 'arm')
 endif
 
 REAL_DIST_DIR:=$(DIST_DIR)/$(CONFIGURATION)/lib/$(MACHINE)

--- a/config.mk
+++ b/config.mk
@@ -9,6 +9,8 @@ ifndef MACHINE
         MACHINE:=x64
     else ifeq ($(MACHINE_RAW),amd64)
         MACHINE:=x64
+    else ifeq ($(MACHINE_RAW),armv7l)
+        MACHINE:=arm
     else
         MACHINE:=x86
     endif

--- a/wrapper/python/mbientlab/metawear/core.py
+++ b/wrapper/python/mbientlab/metawear/core.py
@@ -40,7 +40,7 @@ class DataTypeId:
 # Python wrapper for the MblMwMessage struct
 class Data(Structure):
     _fields_= [
-        ("epoch", c_long),
+        ("epoch", c_longlong),
         ("value", c_void_p),
         ("type_id", c_int),
         ("length", c_ubyte)
@@ -48,10 +48,10 @@ class Data(Structure):
 
 class GattCharacteristic(Structure):
     _fields_= [
-        ("service_uuid_high", c_ulong),
-        ("service_uuid_low", c_ulong),
-        ("uuid_high", c_ulong),
-        ("uuid_low", c_ulong)
+        ("service_uuid_high", c_ulonglong),
+        ("service_uuid_low", c_ulonglong),
+        ("uuid_high", c_ulonglong),
+        ("uuid_low", c_ulonglong)
     ]
 
 Fn_ByteArray= CFUNCTYPE(None, POINTER(c_ubyte), c_ubyte)
@@ -61,7 +61,7 @@ Fn_DataPtr= CFUNCTYPE(None, POINTER(Data))
 Fn_VoidPtr_GattCharPtr_ByteArray= CFUNCTYPE(None, c_void_p, POINTER(GattCharacteristic), POINTER(c_ubyte), c_ubyte)
 Fn_VoidPtr_GattCharPtr= CFUNCTYPE(None, c_void_p, POINTER(GattCharacteristic))
 Fn_Uint_Uint= CFUNCTYPE(None, c_uint, c_uint)
-Fn_Ubyte_Long_ByteArray= CFUNCTYPE(None, c_ubyte, c_long, POINTER(c_ubyte), c_ubyte)
+Fn_Ubyte_Long_ByteArray= CFUNCTYPE(None, c_ubyte, c_longlong, POINTER(c_ubyte), c_ubyte)
 
 # UUIDs for the MetaWear gatt services and characteristics
 class Gatt:

